contracts/evm/contracts/adapters/ClusterConnection.sol

```diff
@@ -0,0 +1,255 @@
+        string calldata networkId,
+        uint256 messageFee,
+        uint256 responseFee
+    ) external onlyAdmin {
+        messageFees[networkId] = messageFee;
+        responseFees[networkId] = responseFee;
+    }
+
+    /**
+     @notice Gets the fee to the target network
+    @param to String Network Id of target chain
+    @param response Boolean ( Whether the responding fee is included )
+    @return fee Integer (The fee of sending a message to a given destination network )
+    */
+    function getFee(
+        string memory to,
+        bool response
+    ) external view returns (uint256 fee) {
+        uint256 messageFee = messageFees[to];
+        if (response == true) {
+            uint256 responseFee = responseFees[to];
+            return messageFee + responseFee;
+        }
+        return messageFee;
+    }
+
+    /**
+     @notice Sends the message to a specific network.
+     @param sn : positive for two-way message, zero for one-way message, negative for response
+     @param to  String ( Network Id of destination network )
+     @param svc String ( name of the service )
+     @param sn  Integer ( serial number of the xcall message )
+     @param _msg Bytes ( serialized bytes of Service Message )
+     */
+    function sendMessage(
+        string calldata to,
+        string calldata svc,
+        int256 sn,
+        bytes calldata _msg
+    ) external payable override {
+        require(msg.sender == xCall, "Only Xcall can call sendMessage");
+        uint256 fee;
+        if (sn > 0) {
+            fee = this.getFee(to, true);
+        } else if (sn == 0) {
+            fee = this.getFee(to, false);
+        }
+        require(msg.value >= fee, "Fee is not Sufficient");
+        connSn++;
+        emit Message(to, connSn, _msg);
+    }
+
+    /**
+     @notice Sends the message to a xCall.
+     @param srcNetwork  String ( Network Id )
+     @param _connSn Integer ( connection message sn )
+     @param _msg Bytes ( serialized bytes of Service Message )
+     */
+    function recvMessageWithSignatures(
+        string memory srcNetwork,
+        uint256 _connSn,
+        bytes calldata _msg,
+        bytes[] calldata _signedMessages
+    ) public onlyAdmin {
+        require(_signedMessages.length > 0, "No signatures provided");
+        require(_signedMessages.length >= reqValidatorCnt, "Not enough signatures passed");
+        bytes32 messageHash = keccak256(_msg);
+        uint signerCount = 0;
+        address[] memory collectedSigners = new address[](_signedMessages.length);
+        for (uint i = 0; i < _signedMessages.length; i++) {
+            address signer = recoverSigner(messageHash, _signedMessages[i]);
+            require(signer != address(0), "Invalid signature");
+            if (!isValidatorProcessed(collectedSigners, signer)){
+                collectedSigners[signerCount] = signer;
+                signerCount++;
+            }
+        }
+
+        if (signerCount >= reqValidatorCnt) {
```

Comment:
If not enough, reverting the transaction could be better approach

Reply:
updated.